src/flb_plugin_proxy.c

```diff
@@ -50,15 +50,6 @@ static void proxy_cb_flush(struct flb_event_chunk *event_chunk,
     (void) i_ins;
     (void) config;
 
-    /* To prevent flush callback executions, we need to check the
-     * status of hot-reloading. The actual problem is: we don't have
-     * pause procedure/mechanism for output plugin. For now, we just halt the
-     * flush callback here during hot-reloading is in progress. */
-    if (config->shutdown_by_hot_reloading == FLB_TRUE) {
-        flb_trace("[GO] hot-reloading is in progress. Retry flushing");
-        FLB_OUTPUT_RETURN(FLB_RETRY);
-    }
-
```

Comment:
When we wrote this condition, there is no pause callback bindings to the external Golang plugins.
Nowadays, there is flb_proxy_input_cb_pause function to pause flushing for Golang plugins.
So, I'd like to suggest to verify the existence of pause callback binding in Golang plugins and registered in this proxy.
Just removing this condition is dangerous and having a possibility to lose logs during hot-reloading.

Reply:
@cosmo0920, thanks for looking into this. But I think there is some misunderstanding. The problem I am facing is with **output** go plugins. `flb_proxy_input_cb_pause` as the name suggests, it is for **input** go plugins. So yes, it makes sense to pause input go plugins, the same way fluent-bit also pauses any other internal `c` input plugins using `flb_input_collector_pause`. But there is no such pause mechanism for internal `c` **output** plugins because I believe there is no need to pause output plugins during reload. 

The code I am removing is from `proxy_cb_flush` function which is used only for output go plugins. If I understand correctly, [calyptia/plugin](https://github.com/chronosphereio/calyptia-plugin) is used for building input go plugins. And since `proxy_cb_flush` will never be called from any input plugins, it should not affect calyptia/plugin](https://github.com/chronosphereio/calyptia-plugin). 

As you can see [here](https://github.com/fluent/fluent-bit/blob/b785118ea63d35b440c816600566794c3d30cec0/src/flb_plugin_proxy.c#L413), `in->cb_flush_buf` is specifically set to `NULL` for `flb_proxy_register_input` because input plugins will never have flush, input plugins have `in->cb_collect = flb_proxy_input_cb_collect;` so this change only affects output go plugins. 

So from what I understand, in simple terms, during a hot reload, fluent-bit is already pausing all the input plugins including the internal `c` and external `go` plugins using `in->cb_pause = flb_proxy_input_cb_pause;`. Once all the input plugins are paused, fluent-bit waits for the pending chunks to be flushed by output plugins. The inbuilt `c` output plugins have no pause mechanism so they continue to flush out remaining chunks which just takes a few seconds and then fluent-bit continues with the reload process. But if you are using external go output plugin, this code doesn't let the output go plugins flush out the remaining chunks, instead it just keeps telling fluent-bit engine to `FLB_RETRY` and in turn fluent-bit engine keep retrying to flush the chunks. This takes several minutes of back and forth until the retries are exhausted and the pending chunks are dropped. After this, the reload process continues. So we not just lost the pending chunks, we also delayed the config reload by several minutes and because of that we also lose/delay new logs that required the new config.

In short, fluent-bit is already pausing all the input plugins, so new logs are not being ingested, so why not let the external go **output** plugins flush out the remaining chunks. Why are we trying to just pause the flushing for external go **output** plugins but not for internal `c` **output** plugins. 

So I still believe that this code is not required and removing that should not be dangerous. As I explained here, we are actually losing a lot of logs due to current behaviour. You can easily replicate the issue as mentioned in #9733.

Sorry for the long reply 😄 , but I hope it clarifies the problem. Please let me know if I am wrong and if you still think this code is required. I am happy to contribute more hours to fixing this problem because with the current behaviour we really can't use any external go **output** plugins.

Reply:
> So from what I understand, in simple terms, during a hot reload, fluent-bit is already pausing all the input plugins including the internal `c` using `flb_input_collector_pause` and external `go` plugins using `in->cb_pause = flb_proxy_input_cb_pause;`. Once all the input plugins are paused, fluent-bit waits for the pending chunks to be flushed by output plugins. The inbuilt `c` output plugins have no pause mechanism so they continue to flush out remaining chunks which just takes a few seconds and then fluent-bit continues with the reload process. But if you are using external go output plugin, this code doesn't let the output go plugins flush out the remaining chunks, instead it just keeps telling fluent-bit engine to `FLB_RETRY` and in turn fluent-bit engine keep retrying to flush the chunks. This takes several minutes of back and forth until the retries are exhausted and the pending chunks are dropped. After this, the reload process continues. So we not just lost the pending chunks, we also delayed the config reload by several minutes and because of that we also lose/delay new logs that required the new config.

No. It's partially correct but there is unprocessed input chunks are still existing when running tasks. We need to deny to mark the unprocessed chunks as retry needed. Once the chunks are marked as processed, they are completely gone. There is no recover way to resubmit into pipelines.

Reply:
> We need to mark the unprocessed chunks as retry needed.

I assume that by **unprocessed chunks**, you mean the data collected or read by input plugins but not yet flushed by an output plugin. Could you please confirm if this definition of **unprocessed chunks** is accurate? My following statements are based on this assumption.

If you mark the unprocessed chunks as "retry," the reload process gets delayed by several minutes. As seen in [this code](https://github.com/fluent/fluent-bit/blob/671cdcd467d88dc8fc56657e8f7a4e90e9e9ddda/src/flb_engine.c#L959-L972), `[engine] service will shutdown when all remaining tasks are flushed.` Since this approach prevents Go **output** plugins from flushing these chunks, the engine keeps retrying to flush the unprocessed chunks, eventually exhausting retries and dropping the chunks permanently. This behavior ensures that unprocessed chunks are always lost.

> Once the chunks are marked as processed, they are completely gone. There is no recoverable way to resubmit them into pipelines.

I don’t fully understand this statement. We are not just marking chunks as processed; we are letting the Go **output** plugins flush these chunks as they normally do. During hot reload, the Go **output** plugins are neither exited nor unregistered. They continue to function until all unprocessed chunks are flushed. Only then does Fluent Bit call `FLBPluginExit` / `FLBPluginExitCtx`, followed by `FLBPluginUnregister`. How would unprocessed chunks be lost in this scenario?

We’ve tested this thoroughly with our custom BigQuery Go **output** plugin. By removing this code, we consistently see all logs, including those in pending tasks, successfully delivered to BigQuery without any loss. However, with this code in place, we are guaranteed to lose all logs in pending tasks.

To provide further clarity, I will add log lines to explain what is happening.

---

### Current Behavior Without My Fix

To replicate the issue, follow the steps mentioned in #9733.  

Build the Go plugin and run `docker-compose up` to start Fluent Bit alongside a log-generating application.  
Here’s a trimmed version of the logs to highlight the relevant parts:

```log
fluent-bit-1     | 2024/12/18 05:58:28 [multiinstance] Register called
fluent-bit-1     | Fluent Bit v3.2.2
fluent-bit-1     | * Copyright (C) 2015-2024 The Fluent Bit Authors
fluent-bit-1     | * Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
fluent-bit-1     | * https://fluentbit.io
fluent-bit-1     | 
fluent-bit-1     | ______ _                  _    ______ _ _           _____  _____ 
fluent-bit-1     | |  ___| |                | |   | ___ (_) |         |____ |/ __  \
fluent-bit-1     | | |_  | |_   _  ___ _ __ | |_  | |_/ /_| |_  __   __   / /`' / /'
fluent-bit-1     | |  _| | | | | |/ _ \ '_ \| __| | ___ \ | __| \ \ / /   \ \  / /  
fluent-bit-1     | | |   | | |_| |  __/ | | | |_  | |_/ / | |_   \ V /.___/ /./ /___
fluent-bit-1     | \_|   |_|\__,_|\___|_| |_|\__| \____/|_|\__|   \_/ \____(_)_____/
fluent-bit-1     | 
fluent-bit-1     | 
fluent-bit-1     | [2024/12/18 05:58:28] [ info] [fluent bit] version=3.2.2, commit=a59c867924, pid=1
fluent-bit-1     | [2024/12/18 05:58:28] [ info] [storage] ver=1.5.2, type=memory, sync=normal, checksum=off, max_chunks_up=128
fluent-bit-1     | [2024/12/18 05:58:28] [ info] [simd    ] disabled
fluent-bit-1     | [2024/12/18 05:58:28] [ info] [cmetrics] version=0.9.9
fluent-bit-1     | [2024/12/18 05:58:28] [ info] [ctraces ] version=0.5.7
fluent-bit-1     | [2024/12/18 05:58:28] [ info] [input:tail:fluent-tail-input] initializing
fluent-bit-1     | [2024/12/18 05:58:28] [ info] [input:tail:fluent-tail-input] storage_strategy='memory' (memory only)
fluent-bit-1     | [2024/12/18 05:58:28] [error] [input:tail:fluent-tail-input] parser 'docker' is not registered
fluent-bit-1     | 2024/12/18 05:58:28 [multiinstance] id = "dummy_metrics"
fluent-bit-1     | 2024/12/18 05:58:28 [multiinstance] id = "dummy_metrics"
fluent-bit-1     | 2024/12/18 05:58:28 [multiinstance] id = "dummy_metrics"
fluent-bit-1     | [2024/12/18 05:58:28] [ info] [http_server] listen iface=0.0.0.0 tcp_port=2020
fluent-bit-1     | [2024/12/18 05:58:28] [ info] [sp] stream processor started
fluent-bit-1     | [2024/12/18 05:58:28] [ info] [input:tail:fluent-tail-input] inotify_fs_add(): inode=45 watch_fd=1 name=/var/log/test.log
fluent-bit-1     | 2024/12/18 05:58:28 [multiinstance] Flush called for id: dummy_metrics
fluent-bit-1     | 2024/12/18 05:58:28 [multiinstance] Flush called for id: dummy_metrics
fluent-bit-1     | 2024/12/18 05:58:28 [multiinstance] Flush called for id: dummy_metrics
...
```

Now, execute the following command from another terminal:

```bash
curl -X POST -d '{}' localhost:2020/api/v2/reload
```

As soon as you send the `SIGHUP` signal to Fluent Bit, the following happens:

```log
...
fluent-bit-1     | 2024/12/18 05:59:15 [multiinstance] Flush called for id: dummy_metrics
fluent-bit-1     | [2024/12/18 05:59:16] [engine] caught signal (SIGHUP)
fluent-bit-1     | [2024/12/18 05:59:16] [ info] reloading instance pid=1 tid=0xffff9de28020
fluent-bit-1     | [2024/12/18 05:59:16] [ info] [reload] slist externals /etc/out_multiinstance.so
fluent-bit-1     | 2024/12/18 05:59:16 [multiinstance] Register called
fluent-bit-1     | [2024/12/18 05:59:16] [ info] [reload] stop everything of the old context
fluent-bit-1     | [2024/12/18 05:59:16] [ warn] [engine] service will shutdown when all remaining tasks are flushed
fluent-bit-1     | [2024/12/18 05:59:16] [ info] [input] pausing fluent-tail-input
fluent-bit-1     | [2024/12/18 05:59:16] [ warn] [engine] failed to flush chunk '1-1734501555.930145462.flb', retry in 1 seconds: task_id=0, input=fluent-tail-input > output=new_alias (out_id=0)
fluent-bit-1     | [2024/12/18 05:59:16] [ warn] [engine] failed to flush chunk '1-1734501555.930145462.flb', retry in 1 seconds: task_id=0, input=fluent-tail-input > output=new_alias_2 (out_id=1)
fluent-bit-1     | [2024/12/18 05:59:16] [ warn] [engine] failed to flush chunk '1-1734501555.930145462.flb', retry in 1 seconds: task_id=0, input=fluent-tail-input > output=new_alias_3 (out_id=2)
fluent-bit-1     | [2024/12/18 05:59:16] [ info] [task] tail/fluent-tail-input has 1 pending task(s):
fluent-bit-1     | [2024/12/18 05:59:16] [ info] [task]   task_id=0 still running on route(s): multiinstance/new_alias multiinstance/new_alias_2 multiinstance/new_alias_3 
fluent-bit-1     | [2024/12/18 05:59:16] [ info] [input] pausing fluent-tail-input
fluent-bit-1     | [2024/12/18 05:59:16] [ warn] [engine] failed to flush chunk '1-1734501555.930145462.flb', retry in 1 seconds: task_id=0, input=fluent-tail-input > output=new_alias (out_id=0)
fluent-bit-1     | [2024/12/18 05:59:16] [ warn] [engine] failed to flush chunk '1-1734501555.930145462.flb', retry in 1 seconds: task_id=0, input=fluent-tail-input > output=new_alias_2 (out_id=1)
fluent-bit-1     | [2024/12/18 05:59:16] [ warn] [engine] failed to flush chunk '1-1734501555.930145462.flb', retry in 1 seconds: task_id=0, input=fluent-tail-input > output=new_alias_3 (out_id=2)
fluent-bit-1     | [2024/12/18 05:59:17] [ info] [input] pausing fluent-tail-input
fluent-bit-1     | [2024/12/18 05:59:17] [ warn] [engine] failed to flush chunk '1-1734501555.930145462.flb', retry in 1 seconds: task_id=0, input=fluent-tail-input > output=new_alias (out_id=0)
fluent-bit-1     | [2024/12/18 05:59:17] [ warn] [engine] failed to flush chunk '1-1734501555.930145462.flb', retry in 1 seconds: task_id=0, input=fluent-tail-input > output=new_alias_2 (out_id=1)
fluent-bit-1     | [2024/12/18 05:59:17] [ warn] [engine] failed to flush chunk '1-1734501555.930145462.flb', retry in 1 seconds: task_id=0, input=fluent-tail-input > output=new_alias_3 (out_id=2)
fluent-bit-1     | [2024/12/18 05:59:18] [ info] [input] pausing fluent-tail-input
...
```

After receiving the `SIGHUP` signal, Fluent Bit logs: `service will shutdown when all remaining tasks are flushed.` Then, it repeatedly logs:  
`failed to flush chunk '1-1734501555.930145462.flb', retry in 1 seconds: task_id=0, input=fluent-tail-input > output=new_alias (out_id=0).`  
This retry process repeats 10 times (as retry_limit is set to 10 in the output configuration).

After exhausting the retries, Fluent Bit drops the unprocessed chunks and logs:  
```log
...
fluent-bit-1     | [2024/12/18 05:59:25] [error] [engine] chunk '1-1734501555.930145462.flb' cannot be retried: task_id=0, input=fluent-tail-input > output=new_alias_2
fluent-bit-1     | [2024/12/18 05:59:25] [error] [engine] chunk '1-1734501555.930145462.flb' cannot be retried: task_id=0, input=fluent-tail-input > output=new_alias
fluent-bit-1     | [2024/12/18 05:59:25] [error] [engine] chunk '1-1734501555.930145462.flb' cannot be retried: task_id=0, input=fluent-tail-input > output=new_alias_3
fluent-bit-1     | [2024/12/18 05:59:26] [ info] [engine] service has stopped (0 pending tasks)
fluent-bit-1     | [2024/12/18 05:59:26] [ info] [input] pausing fluent-tail-input
fluent-bit-1     | 2024/12/18 05:59:26 [multiinstance] Exit called for id: dummy_metrics
fluent-bit-1     | 2024/12/18 05:59:26 [multiinstance] Exit called for id: dummy_metrics
fluent-bit-1     | 2024/12/18 05:59:26 [multiinstance] Exit called for id: dummy_metrics
fluent-bit-1     | [2024/12/18 05:59:26] [ info] [input:tail:fluent-tail-input] inotify_fs_remove(): inode=45 watch_fd=1
fluent-bit-1     | 2024/12/18 05:59:26 [multiinstance] Unregister called
fluent-bit-1     | [2024/12/18 05:59:26] [ info] [reload] start everything
fluent-bit-1     | [2024/12/18 05:59:26] [ info] [fluent bit] version=3.2.2, commit=a59c867924, pid=1
fluent-bit-1     | [2024/12/18 05:59:26] [ info] [storage] ver=1.5.2, type=memory, sync=normal, checksum=off, max_chunks_up=128
fluent-bit-1     | [2024/12/18 05:59:26] [ info] [simd    ] disabled
fluent-bit-1     | [2024/12/18 05:59:26] [ info] [cmetrics] version=0.9.9
fluent-bit-1     | [2024/12/18 05:59:26] [ info] [ctraces ] version=0.5.7
fluent-bit-1     | [2024/12/18 05:59:26] [ info] [input:tail:fluent-tail-input] initializing
fluent-bit-1     | [2024/12/18 05:59:26] [ info] [input:tail:fluent-tail-input] storage_strategy='memory' (memory only)
fluent-bit-1     | [2024/12/18 05:59:26] [error] [input:tail:fluent-tail-input] parser 'docker' is not registered
fluent-bit-1     | 2024/12/18 05:59:26 [multiinstance] id = "dummy_metrics"
fluent-bit-1     | 2024/12/18 05:59:26 [multiinstance] id = "dummy_metrics"
...
```
It then logs: `[engine] service has stopped (0 pending tasks).` After a 10-second delay (the 10-second delay is rather small because I replicated the issue in a very simple configuration, in our actual setup with a complex pipeline with high load of incoming logs the delay is 5-10 minutes), it resumes the reload process, calls `FLBPluginExitCtx` for the Go **output** plugins (evident from `[multiinstance] Exit called for id: dummy_metrics`) and then `FLBPluginUnregister`.

This behavior confirms that Go **output** plugins are not exited or unregistered when a hot reload starts. Fluent Bit waits for unprocessed chunks to flush before exiting and unregistering the plugins. It also proves that the current implementation guarantees unprocessed chunks in pending tasks will be lost when using Go **output** plugins.

I have attached the full log file for reference.

---

### Questions

1. Why is this flushing behavior enforced for Go **output** plugins but not for internal `C` plugins? If we need to mark unprocessed chunks as "retry needed", why isn’t this applied to internal `C` plugins as well?
   
2. From our discussions, you believe that this code is needed to prevent log loss during hot reload. However, the current behavior seems to ensure the opposite—it guarantees the permanent loss of unprocessed logs. Can you please explain why this code is considered necessary?

Answering these questions will help me better understand your perspective and why you believe this code is required.

Reply:
The full log file.
[current-behaviour.log](https://github.com/user-attachments/files/18176362/current-behaviour.log)

Reply:
> 1. Why is this flushing behavior enforced for Go **output** plugins but not for internal `C` plugins? If we need to mark unprocessed chunks as "retry needed", why isn’t this applied to internal `C` plugins as well?

This is because internal C plugins have pause callbacks. For Golang plugins, there is no loading and executing pause callbacks.
This is why there is first-aid for blocking to flush unprocessed chunks.



> 2\. From our discussions, you believe that this code is needed to prevent log loss during hot reload. However, the current behavior seems to ensure the opposite—it guarantees the permanent loss of unprocessed logs. Can you please explain why this code is considered necessary?

We intended to prevent losses of unprocessed chunks with filesystem storage buffering for chunks. I looked your repro but there is no usage of filesystem type of storage.
Without storage, Fluent Bit mostly loses their data during hot-reloading and also terminating.